File: packages/@emulators/core/src/store.ts
export interface Entity {
  id: number;
  created_at: string;
  updated_at: string;
}

export type InsertInput<T extends Entity> = Omit<T, "id" | "created_at" | "updated_at"> & { id?: number };

export type FilterFn<T> = (item: T) => boolean;
export type SortFn<T> = (a: T, b: T) => number;

export interface QueryOptions<T> {
  filter?: FilterFn<T>;
  sort?: SortFn<T>;
  page?: number;
  per_page?: number;
}

export interface PaginatedResult<T> {
  items: T[];
  total_count: number;
  page: number;
  per_page: number;
  has_next: boolean;
  has_prev: boolean;
}

export interface CollectionSnapshot<T extends Entity = Entity> {
  items: T[];
  autoId: number;
  indexFields: string[];
}

export interface StoreSnapshot {
  collections: Record<string, CollectionSnapshot>;
  data: Record<string, unknown>;
}

export function serializeValue(value: unknown): unknown {
  if (value instanceof Map) {
    return { __type: "Map" as const, entries: [...value.entries()].map(([k, v]) => [k, serializeValue(v)]) };
  }
  if (value instanceof Set) {
    return { __type: "Set" as const, values: [...value.values()] };
  }
  return value;
}

export function deserializeValue(value: unknown): unknown {
  if (value !== null && typeof value === "object" && "__type" in value) {
    const tagged = value as Record<string, unknown>;
    if (tagged.__type === "Map") {
      const entries = tagged.entries as [unknown, unknown][];
      return new Map(entries.map(([k, v]) => [k, deserializeValue(v)]));
    }
    if (tagged.__type === "Set") {
      return new Set(tagged.values as unknown[]);
    }
  }
  return value;
}

export class Collection<T extends Entity> {
  private items = new Map<number, T>();
  private indexes = new Map<string, Map<string | number, Set<number>>>();
  private autoId = 1;
  readonly fieldNames: string[];

  constructor(private indexFields: (keyof T)[] = []) {
    this.fieldNames = indexFields.map(String).sort();
    for (const field of indexFields) {
      this.indexes.set(String(field), new Map());
    }
  }

  private addToIndex(item: T): void {
    for (const field of this.indexFields) {
      const value = item[field];
      if (value === undefined || value === null) continue;
      const indexMap = this.indexes.get(String(field))!;
      const key = String(value);
      if (!indexMap.has(key)) {
        indexMap.set(key, new Set());
      }
      indexMap.get(key)!.add(item.id);
    }
  }

  private removeFromIndex(item: T): void {
    for (const field of this.indexFields) {
      const value = item[field];
      if (value === undefined || value === null) continue;
      const indexMap = this.indexes.get(String(field))!;
      const key = String(value);
      indexMap.get(key)?.delete(item.id);
    }
  }

  insert(data: InsertInput<T>): T {
    const now = new Date().toISOString();
    const explicitId = data.id != null && data.id > 0 ? data.id : undefined;
    const id = explicitId ?? this.autoId++;
    if (id >= this.autoId) {
      this.autoId = id + 1;
    }
    const item = {
      ...data,
      id,
      created_at: now,
      updated_at: now,
    } as unknown as T;
    this.items.set(id, item);
    this.addToIndex(item);
    return item;
  }

  get(id: number): T | undefined {
    return this.items.get(id);
  }

  findBy(field: keyof T, value: T[keyof T] | string | number): T[] {
    if (this.indexes.has(String(field))) {
      const ids = this.indexes.get(String(field))!.get(String(value));
      if (!ids) return [];
      return Array.from(ids)
        .map((id) => this.items.get(id)!)
        .filter(Boolean);
    }
    return this.all().filter((item) => item[field] === value);
  }

  findOneBy(field: keyof T, value: T[keyof T] | string | number): T | undefined {
    return this.findBy(field, value)[0];
  }

  update(id: number, data: Partial<T>): T | undefined {
    const existing = this.items.get(id);
    if (!existing) return undefined;
    this.removeFromIndex(existing);
    const updated = {
      ...existing,
      ...data,
      id,
      updated_at: new Date().toISOString(),
    } as T;
    this.items.set(id, updated);
    this.addToIndex(updated);
    return updated;
  }

  delete(id: number): boolean {
    const existing = this.items.get(id);
    if (!existing) return false;
    this.removeFromIndex(existing);
    return this.items.delete(id);
  }

  all(): T[] {
    return Array.from(this.items.values());
  }

  query(options: QueryOptions<T> = {}): PaginatedResult<T> {
    let results = this.all();

    if (options.filter) {
      results = results.filter(options.filter);
    }

    const total_count = results.length;

    if (options.sort) {
      results.sort(options.sort);
    }

    const page = options.page ?? 1;
    const per_page = Math.min(options.per_page ?? 30, 100);
    const start = (page - 1) * per_page;
    const paged = results.slice(start, start + per_page);

    return {
      items: paged,
      total_count,
      page,
      per_page,
      has_next: start + per_page < total_count,
      has_prev: page > 1,
    };
  }

  count(filter?: FilterFn<T>): number {
    if (!filter) return this.items.size;
    return this.all().filter(filter).length;
  }

  clear(): void {
    this.items.clear();
    for (const indexMap of this.indexes.values()) {
      indexMap.clear();
    }
    this.autoId = 1;
  }

  snapshot(): CollectionSnapshot<T> {
    return {
      items: this.all(),
      autoId: this.autoId,
      indexFields: this.fieldNames,
    };
  }

  restore(snap: CollectionSnapshot<T>): void {
    this.clear();
    this.autoId = snap.autoId;
    for (const item of snap.items) {
      this.items.set(item.id, item);
      this.addToIndex(item);
    }
  }
}

export class Store {
  private collections = new Map<string, Collection<any>>();
  private _data = new Map<string, unknown>();

  collection<T extends Entity>(name: string, indexFields: (keyof T)[] = []): Collection<T> {
    const existing = this.collections.get(name);
    if (existing) {
      if (indexFields.length > 0) {
        const requested = indexFields.map(String).sort();
        if (existing.fieldNames.length !== requested.length || existing.fieldNames.some((f, i) => f !== requested[i])) {
          throw new Error(
            `Collection "${name}" already exists with indexes [${existing.fieldNames}] but was requested with [${requested}]`,
          );
        }
      }
      return existing as Collection<T>;
    }
    const col = new Collection<T>(indexFields);
    this.collections.set(name, col);
    return col;
  }

  getData<V>(key: string): V | undefined {
    return this._data.get(key) as V | undefined;
  }

  setData<V>(key: string, value: V): void {
    this._data.set(key, value);
  }

  reset(): void {
    for (const collection of this.collections.values()) {
      collection.clear();
    }
    this._data.clear();
  }

  snapshot(): StoreSnapshot {
    const collections: Record<string, CollectionSnapshot> = {};
    for (const [name, col] of this.collections) {
      collections[name] = col.snapshot();
    }
    const data: Record<string, unknown> = {};
    for (const [key, value] of this._data) {
      data[key] = serializeValue(value);
    }
    return { collections, data };
  }

  restore(snap: StoreSnapshot): void {
    const snapshotNames = new Set(Object.keys(snap.collections));
    for (const name of this.collections.keys()) {
      if (!snapshotNames.has(name)) {
        this.collections.delete(name);
      }
    }
    for (const [name, colSnap] of Object.entries(snap.collections)) {
      const indexFields = colSnap.indexFields as (keyof Entity)[];
      const col = this.collection(name, indexFields);
      col.restore(colSnap as CollectionSnapshot<any>);
    }
    this._data.clear();
    for (const [key, value] of Object.entries(snap.data)) {
      this._data.set(key, deserializeValue(value));
    }
  }
}

File: internal/core/store/store.go
package store

import (
	"encoding/json"
	"fmt"
	"sort"
	"sync"
	"time"
)

type Record map[string]any

type CollectionSnapshot struct {
	Items       []Record `json:"items"`
	AutoID      int      `json:"autoId"`
	IndexFields []string `json:"indexFields"`
}

type StoreSnapshot struct {
	Collections map[string]CollectionSnapshot `json:"collections"`
	Data        map[string]any                `json:"data"`
}

type Store struct {
	mu          sync.RWMutex
	collections map[string]*Collection
	data        map[string]any
}

func New() *Store {
	return &Store{
		collections: map[string]*Collection{},
		data:        map[string]any{},
	}
}

func (s *Store) Collection(name string, indexFields ...string) (*Collection, error) {
	requested := normalizeFields(indexFields)

	s.mu.Lock()
	defer s.mu.Unlock()

	if existing, ok := s.collections[name]; ok {
		if len(requested) > 0 && !sameStrings(existing.IndexFields(), requested) {
			return nil, fmt.Errorf("collection %q already exists with indexes %v but was requested with %v", name, existing.IndexFields(), requested)
		}
		return existing, nil
	}

	collection := newCollection(requested)
	s.collections[name] = collection
	return collection, nil
}

func (s *Store) MustCollection(name string, indexFields ...string) *Collection {
	collection, err := s.Collection(name, indexFields...)
	if err != nil {
		panic(err)
	}
	return collection
}

func (s *Store) SetData(key string, value any) {
	s.mu.Lock()
	defer s.mu.Unlock()
	s.data[key] = cloneValue(value)
}

func (s *Store) GetData(key string) (any, bool) {
	s.mu.RLock()
	defer s.mu.RUnlock()
	value, ok := s.data[key]
	if !ok {
		return nil, false
	}
	return cloneValue(value), true
}

func (s *Store) Reset() {
	s.mu.Lock()
	collections := make([]*Collection, 0, len(s.collections))
	for _, collection := range s.collections {
		collections = append(collections, collection)
	}
	s.data = map[string]any{}
	s.mu.Unlock()

	for _, collection := range collections {
		collection.Clear()
	}
}

func (s *Store) Snapshot() StoreSnapshot {
	s.mu.RLock()
	names := make([]string, 0, len(s.collections))
	for name := range s.collections {
		names = append(names, name)
	}
	sort.Strings(names)

	collections := make(map[string]*Collection, len(s.collections))
	for _, name := range names {
		collections[name] = s.collections[name]
	}

	data := make(map[string]any, len(s.data))
	for key, value := range s.data {
		data[key] = cloneValue(value)
	}
	s.mu.RUnlock()

	snapshots := make(map[string]CollectionSnapshot, len(collections))
	for _, name := range names {
		snapshots[name] = collections[name].Snapshot()
	}

	return StoreSnapshot{Collections: snapshots, Data: data}
}

func (s *Store) Restore(snapshot StoreSnapshot) error {
	s.mu.Lock()
	defer s.mu.Unlock()

	for name := range s.collections {
		if _, ok := snapshot.Collections[name]; !ok {
			delete(s.collections, name)
		}
	}

	for name, collectionSnapshot := range snapshot.Collections {
		fields := normalizeFields(collectionSnapshot.IndexFields)
		collection, ok := s.collections[name]
		if !ok || !sameStrings(collection.IndexFields(), fields) {
			collection = newCollection(fields)
			s.collections[name] = collection
		}
		if err := collection.Restore(collectionSnapshot); err != nil {
			return fmt.Errorf("restore collection %q: %w", name, err)
		}
	}

	s.data = make(map[string]any, len(snapshot.Data))
	for key, value := range snapshot.Data {
		s.data[key] = cloneValue(value)
	}
	return nil
}

func (s *Store) MarshalSnapshot() ([]byte, error) {
	return json.MarshalIndent(s.Snapshot(), "", "  ")
}

func (s *Store) RestoreJSON(raw []byte) error {
	var snapshot StoreSnapshot
	if err := json.Unmarshal(raw, &snapshot); err != nil {
		return err
	}
	return s.Restore(snapshot)
}

type Collection struct {
	mu          sync.RWMutex
	items       map[int]Record
	indexes     map[string]map[string]map[int]struct{}
	indexFields []string
	autoID      int
}

func newCollection(indexFields []string) *Collection {
	indexes := make(map[string]map[string]map[int]struct{}, len(indexFields))
	for _, field := range indexFields {
		indexes[field] = map[string]map[int]struct{}{}
	}
	return &Collection{
		items:       map[int]Record{},
		indexes:     indexes,
		indexFields: append([]string(nil), indexFields...),
		autoID:      1,
	}
}

func (c *Collection) Insert(record Record) Record {
	c.mu.Lock()
	defer c.mu.Unlock()

	now := timestamp()
	item := cloneRecord(record)
	id, ok := numericID(item["id"])
	if !ok || id <= 0 {
		id = c.autoID
	}
	if id >= c.autoID {
		c.autoID = id + 1
	}

	if existing, ok := c.items[id]; ok {
		c.removeFromIndexes(existing)
	}

	item["id"] = id
	if _, ok := item["created_at"]; !ok {
		item["created_at"] = now
	}
	item["updated_at"] = now
	c.items[id] = item
	c.addToIndexes(item)
	return cloneRecord(item)
}

func (c *Collection) Get(id int) (Record, bool) {
	c.mu.RLock()
	defer c.mu.RUnlock()
	item, ok := c.items[id]
	if !ok {
		return nil, false
	}
	return cloneRecord(item), true
}

func (c *Collection) Update(id int, patch Record) (Record, bool) {
	c.mu.Lock()
	defer c.mu.Unlock()

	existing, ok := c.items[id]
	if !ok {
		return nil, false
	}
	c.removeFromIndexes(existing)

	updated := cloneRecord(existing)
	for key, value := range patch {
		if key == "id" {
			continue
		}
		updated[key] = cloneValue(value)
	}
	updated["updated_at"] = timestamp()
	c.items[id] = updated
	c.addToIndexes(updated)
	return cloneRecord(updated), true
}

func (c *Collection) Delete(id int) bool {
	c.mu.Lock()
	defer c.mu.Unlock()
	existing, ok := c.items[id]
	if !ok {
		return false
	}
	c.removeFromIndexes(existing)
	delete(c.items, id)
	return true
}

func (c *Collection) All() []Record {
	c.mu.RLock()
	defer c.mu.RUnlock()
	return c.allLocked()
}

func (c *Collection) FindBy(field string, value any) []Record {
	c.mu.RLock()
	defer c.mu.RUnlock()

	key := indexKey(value)
	if index, ok := c.indexes[field]; ok {
		ids := sortedIDs(index[key])
		records := make([]Record, 0, len(ids))
		for _, id := range ids {
			if item, ok := c.items[id]; ok {
				records = append(records, cloneRecord(item))
			}
		}
		return records
	}

	records := make([]Record, 0)
	for _, item := range c.allLocked() {
		if indexKey(item[field]) == key {
			records = append(records, item)
		}
	}
	return records
}

func (c *Collection) Count() int {
	c.mu.RLock()
	defer c.mu.RUnlock()
	return len(c.items)
}

func (c *Collection) Clear() {
	c.mu.Lock()
	defer c.mu.Unlock()
	c.items = map[int]Record{}
	for _, index := range c.indexes {
		for key := range index {
			delete(index, key)
		}
	}
	c.autoID = 1
}

func (c *Collection) Snapshot() CollectionSnapshot {
	c.mu.RLock()
	defer c.mu.RUnlock()
	return CollectionSnapshot{
		Items:       c.allLocked(),
		AutoID:      c.autoID,
		IndexFields: append([]string(nil), c.indexFields...),
	}
}

func (c *Collection) Restore(snapshot CollectionSnapshot) error {
	c.mu.Lock()
	defer c.mu.Unlock()

	fields := normalizeFields(snapshot.IndexFields)
	if !sameStrings(c.indexFields, fields) {
		c.indexFields = fields
		c.indexes = map[string]map[string]map[int]struct{}{}
		for _, field := range fields {
			c.indexes[field] = map[string]map[int]struct{}{}
		}
	}

	c.items = map[int]Record{}
	for _, index := range c.indexes {
		for key := range index {
			delete(index, key)
		}
	}

	c.autoID = snapshot.AutoID
	if c.autoID < 1 {
		c.autoID = 1
	}
	for _, item := range snapshot.Items {
		id, ok := numericID(item["id"])
		if !ok || id <= 0 {
			return fmt.Errorf("record is missing positive numeric id")
		}
		record := cloneRecord(item)
		record["id"] = id
		c.items[id] = record
		c.addToIndexes(record)
		if id >= c.autoID {
			c.autoID = id + 1
		}
	}
	return nil
}

func (c *Collection) IndexFields() []string {
	c.mu.RLock()
	defer c.mu.RUnlock()
	return append([]string(nil), c.indexFields...)
}

func (c *Collection) allLocked() []Record {
	ids := make([]int, 0, len(c.items))
	for id := range c.items {
		ids = append(ids, id)
	}
	sort.Ints(ids)

	records := make([]Record, 0, len(ids))
	for _, id := range ids {
		records = append(records, cloneRecord(c.items[id]))
	}
	return records
}

func (c *Collection) addToIndexes(item Record) {
	id, ok := numericID(item["id"])
	if !ok {
		return
	}
	for field, index := range c.indexes {
		value, ok := item[field]
		if !ok || value == nil {
			continue
		}
		key := indexKey(value)
		if _, ok := index[key]; !ok {
			index[key] = map[int]struct{}{}
		}
		index[key][id] = struct{}{}
	}
}

func (c *Collection) removeFromIndexes(item Record) {
	id, ok := numericID(item["id"])
	if !ok {
		return
	}
	for field, index := range c.indexes {
		value, ok := item[field]
		if !ok || value == nil {
			continue
		}
		key := indexKey(value)
		delete(index[key], id)
		if len(index[key]) == 0 {
			delete(index, key)
		}
	}
}

func normalizeFields(fields []string) []string {
	seen := map[string]bool{}
	normalized := make([]string, 0, len(fields))
	for _, field := range fields {
		if field == "" || seen[field] {
			continue
		}
		seen[field] = true
		normalized = append(normalized, field)
	}
	sort.Strings(normalized)
	return normalized
}

func sameStrings(a []string, b []string) bool {
	if len(a) != len(b) {
		return false
	}
	for i := range a {
		if a[i] != b[i] {
			return false
		}
	}
	return true
}

func sortedIDs(ids map[int]struct{}) []int {
	if len(ids) == 0 {
		return nil
	}
	out := make([]int, 0, len(ids))
	for id := range ids {
		out = append(out, id)
	}
	sort.Ints(out)
	return out
}

func numericID(value any) (int, bool) {
	switch v := value.(type) {
	case int:
		return v, true
	case int8:
		return int(v), true
	case int16:
		return int(v), true
	case int32:
		return int(v), true
	case int64:
		return int(v), true
	case uint:
		return int(v), true
	case uint8:
		return int(v), true
	case uint16:
		return int(v), true
	case uint32:
		return int(v), true
	case uint64:
		return int(v), true
	case float64:
		id := int(v)
		return id, float64(id) == v
	case json.Number:
		id, err := v.Int64()
		return int(id), err == nil
	default:
		return 0, false
	}
}

func indexKey(value any) string {
	return fmt.Sprint(value)
}

func timestamp() string {
	return time.Now().UTC().Format(time.RFC3339Nano)
}

func cloneRecord(record Record) Record {
	if record == nil {
		return Record{}
	}
	out := make(Record, len(record))
	for key, value := range record {
		out[key] = cloneValue(value)
	}
	return out
}

func cloneValue(value any) any {
	switch v := value.(type) {
	case Record:
		return cloneRecord(v)
	case map[string]any:
		out := make(map[string]any, len(v))
		for key, nested := range v {
			out[key] = cloneValue(nested)
		}
		return out
	case []Record:
		out := make([]Record, len(v))
		for i, nested := range v {
			out[i] = cloneRecord(nested)
		}
		return out
	case []map[string]any:
		out := make([]map[string]any, len(v))
		for i, nested := range v {
			out[i] = cloneValue(nested).(map[string]any)
		}
		return out
	case []any:
		out := make([]any, len(v))
		for i, nested := range v {
			out[i] = cloneValue(nested)
		}
		return out
	case []string:
		return append([]string(nil), v...)
	case []int:
		return append([]int(nil), v...)
	default:
		return v
	}
}
